truffle/src/main/java/org/jruby/truffle/core/rubinius/RegexpPrimitiveNodes.java

```diff
@@ -110,48 +114,114 @@ private String formatError(DynamicObject string) {
-            setLastMatch(getContext(), matchData);
-            return matchData;
+        public static RegexpSetLastMatchPrimitiveNode create() {
+            return RegexpPrimitiveNodesFactory.RegexpSetLastMatchPrimitiveNodeFactory.create(null);
         }
 
+        public abstract DynamicObject executeSetLastMatch(Object matchData);
+
         @TruffleBoundary
-        public static void setLastMatch(RubyContext context, Object matchData) {
-            final DynamicObject threadLocals = Layouts.THREAD.getThreadLocals(context.getThreadManager().getCurrentThread());
-            boolean res = threadLocals.set("$~", matchData);
-            assert res;
+        @Specialization(guards = "isSuitableMatchDataType(getContext(), matchData)")
+        public DynamicObject setLastMatchData(DynamicObject matchData) {
+            Frame frame = getContext().getCallStack().getCallerFrameIgnoringSend().getFrame(FrameInstance.FrameAccess.READ_WRITE, true);
+            FrameSlot slot = frame.getFrameDescriptor().findFrameSlot("$~");
+
+            while (slot == null) {
+                final Frame nextFrame = RubyArguments.getDeclarationFrame(frame);
+
+                if (nextFrame == null) {
+                    slot = frame.getFrameDescriptor().addFrameSlot("$~", FrameSlotKind.Object);
+                } else {
+                    slot = nextFrame.getFrameDescriptor().findFrameSlot("$~");
+                    frame = nextFrame;
+                }
+            }
+
+            final Object previousMatchData;
+            try {
+                previousMatchData = frame.getObject(slot);
+
+                if (previousMatchData instanceof ThreadLocalObject) {
+                    final ThreadLocalObject threadLocalObject = (ThreadLocalObject) previousMatchData;
+
+                    threadLocalObject.set(matchData);
+                } else {
+                    frame.setObject(slot, ThreadLocalObject.wrap(getContext(), matchData));
+                }
+            } catch (FrameSlotTypeException e) {
+                throw new IllegalStateException(e);
+            }
+
+            return matchData;
         }
 
     }
 
-    @Primitive(name = "regexp_set_block_last_match")
+    @Primitive(name = "regexp_set_block_last_match", needsSelf = false)
+    @ImportStatic(RegexpPrimitiveNodes.class)
     public static abstract class RegexpSetBlockLastMatchPrimitiveNode extends PrimitiveArrayArgumentsNode {
 
-        @Specialization
-        public DynamicObject setBlockLastMatch(DynamicObject regexpClass) {
-            // TODO CS 7-Mar-15 what does this do?
-            return nil();
+        @TruffleBoundary
+        @Specialization(guards = { "isRubyProc(block)", "isSuitableMatchDataType(getContext(), matchData)" })
+        public Object setBlockLastMatch(DynamicObject block, DynamicObject matchData) {
+
+            Frame callerFrame = Layouts.PROC.getDeclarationFrame(block);
+
+            if (callerFrame == null) {
```

Comment:
What you're saying makes sense, but I did add it for a reason. I need to look through the history in the IDE to see if it was added for a different approach (I churned through several).